services/src/handlers/mod.rs

```diff
@@ -32,20 +38,68 @@ pub async fn handle_rejection(error: Rejection) -> Result<impl Reply, Rejection>
     })
 }
 
-pub fn authenticate<T: UserDB>(
-    user_db: DB<T>,
-) -> impl warp::Filter<Extract = (Session,), Error = warp::Rejection> + Clone {
-    async fn do_authenticate<T: UserDB>(
-        user_db: DB<T>,
-        token: String,
-    ) -> Result<Session, warp::Rejection> {
+fn authenticate<C: Context>(
+    ctx: C,
+) -> impl warp::Filter<Extract = (C,), Error = warp::Rejection> + Clone {
+    async fn do_authenticate<C: Context>(mut ctx: C, token: String) -> Result<C, warp::Rejection> {
         let token = SessionToken::from_str(&token).map_err(|_| warp::reject())?;
+        let user_db = ctx.user_db();
         let db = user_db.read().await;
-        db.session(token).map_err(|_| warp::reject())
+        ctx.set_session(db.session(token).map_err(|_| warp::reject())?);
+        Ok(ctx)
     }
 
     warp::any()
-        .and(warp::any().map(move || Arc::clone(&user_db)))
+        .and(warp::any().map(move || ctx.clone()))
         .and(warp::header::<String>("authorization"))
         .and_then(do_authenticate)
 }
+
+pub trait Context: 'static + Send + Sync + Clone {
+    type UserDB: UserDB;
+    type ProjectDB: ProjectDB;
+    type WorkflowRegistry: WorkflowRegistry;
+
+    fn user_db(&self) -> DB<Self::UserDB>;
+    fn project_db(&self) -> DB<Self::ProjectDB>;
+    fn workflow_registry(&self) -> DB<Self::WorkflowRegistry>;
+    fn session(&self) -> Result<&Session>;
+
+    fn set_session(&mut self, session: Session);
+}
+
+#[derive(Clone, Default)]
+pub struct DefaultContext {
+    user_db: DB<HashMapUserDB>,
+    project_db: DB<HashMapProjectDB>,
+    workflow_registry: DB<HashMapRegistry>,
+    session: Option<Session>,
+}
+
+impl Context for DefaultContext {
+    type UserDB = HashMapUserDB;
+    type ProjectDB = HashMapProjectDB;
+    type WorkflowRegistry = HashMapRegistry;
+
+    fn user_db(&self) -> DB<Self::UserDB> {
+        self.user_db.clone()
+    }
+
+    fn project_db(&self) -> DB<Self::ProjectDB> {
+        self.project_db.clone()
+    }
+
+    fn workflow_registry(&self) -> DB<Self::WorkflowRegistry> {
+        self.workflow_registry.clone()
+    }
```

Comment:
Is it necessary to always clone it or to hand out a reference for some cases?

Reply:
Arc is already kind of a reference. Does it really make sense to hand out a reference to a reference? Maybe it makes more sense to (additionally) provide a method that directly returns the contained Mutex or even the Read/Write guard. Latter one would introduce two methods for each contained value for read/write...

Reply:
I see lots of calls where it is not necessary to clone it.
So you could hand out some read guard with `project_db_ref()` and write guard with `project_db_mut()`.

Reply:
c6abcb51f4bd8e4afa329878a50068a4b73ec59f